0-module-and-module-source.md

```diff
@@ -0,0 +1,371 @@
+and evaluate modules in their particular realm.
+
+## Examples
+
+### Import Kicker
+
+Any dynamic import function is suitable for initializing, linking, and
+evaluating a module instance.
+This necessarily implies advancing all of its transitive dependencies to their
+terminal state or any one into a failed state.
+
+```js
+const source = new ModuleSource(``);
+const instance = new Module(source, {
+  importHook,
+  importMeta: import.meta,
+});
+const namespace = await import(instance);
+```
+
+### Module Idempotency
+
+Since the Module has a bound module namespace exotic object, importing the same
+instance must yield the same result:
+
+```js
+const source = new ModuleSource(``);
+const instance = new Module(source, {
+  importHook,
+  importMeta: import.meta,
+});
+const namespace1 = await import(instance);
+const namespace2 = await import(instance);
+namespace1 === namespace2; // true
+```
+
+### Reusing ModuleSource
+
+Module sources are backed by a shared immutable module source record
+that can be instantiated multiple times, even locally.
+Multiple `Module` instances can share a module source and produce
+separate module namespaces.
+
+```js
+const source = new ModuleSource(``);
+const instance1 = new Module(source, {
+  importHook: importHook1,
+  importMeta: import.meta,
+});
+const instance2 = new Module(source, {
+  importHook: importHook2,
+  importMeta: import.meta,
+}));
+instance1 === instance2; // false
+const namespace1 = await import(instance1);
+const namespace2 = await import(instance2);
+namespace1 === namespace2; // false
+```
+
+### Intersection Semantics with Module Blocks
+
+Proposal: https://github.com/tc39/proposal-js-module-blocks
+
+In relation to module blocks, we can extend the proposal to accommodate both,
+the concept of a module block instance and module block source:
+
+```js
+const instance = module {};
+instance instanceof Module;
+instance.source instanceof ModuleSource;
+const namespace = await import(instance);
+```
+
+To avoid needing a throw-away module-instance in order to get a module source,
+we can extend the syntax:
+
+```js
+const source = static module {};
+source instanceof ModuleSource;
+const instance = new Module(source, {
+  importHook,
+  importMeta: import.meta,
+});
+const namespace = await import(instance);
+```
+
+### Intersection Semantics with deferred execution
+
+The possibility to load the source, and create the instance with the default
+`importHook` and the `import.meta` of the importer, that can be imported at any
+given time, is sufficient:
+
+```js
+import instance from 'module.js' deferred execution syntax;
+instance instanceof Module;
+instance.source instanceof ModuleSource;
+const namespace = await import(instance);
+```
+
+If the goal is to also control the `importHook` and the `importMeta` of the
+importer, then a new syntax can be provided to only get the `ModuleSource`:
+
+```ts
+import source from 'module.js' static source syntax;
```

Comment:
Question: If we have the second, giving us the ModuleSource, does that subsume the need for the first?

If so, and given the `static module {}` syntax above, could this one be

```js
import source from static 'module.js';
```
Isn't this exactly the same issue as module reflection? However it is spelled, doesn't the ability to import the ModuleSource subsume all of these?

Reply:
Yes, importing module sources overlaps with import reflection and importing module instances overlaps with dynamic import. There are many ways that Module Harmony can unfold that cover all the motivating cases, but I’m in favor of separate `module` and `static module` syntaxes for both module blocks and import (by whatever spelling) , to cover both deferred execution and module source reflection.

```js
import module from 'module';
module instancof Module // true, maybe pre-loaded, must not be executed

import static module from 'module';
// module is a module source, albeit `ModuleSource`, `WebAssembly.Module`, or a virtual module source
// maybe virtual module sources are instances of `ModuleSource` at this juncture

(module {}) instanceof Module
(module {}).source instanceof ModuleSource
(static module {}) instanceof ModuleSource
```